### PR TITLE
Updated cookie secrets management

### PR DESCRIPTION
**Pull request by wtoorop**
This PR is in response to issue #196 (and thus also of interest to @anandb-ripencc  )
- The cookie secret file's default location can be set with `--with-cookiesecretsfile=path` with `configure`
- The default cookie secrets file is changed to `{dbdir}/cookiesecrets.txt`
- When no cookie secrets file location is set with the `cookie-secret-file` option in the config file, and the new default location does not exist, cookie secrets will be read from the previous default location `{configdir}/nsd_cookiesecrets.txt`
- A staging cookie secret can also be provided in the config file with the `cookie-staging-secret` option.
- Explicitly configured cookie secrets in config file (with the `cookie-secret` and `cookie-staging-secret` option) take precedence over the ones from the cookie secret file
- All values of DNS cookie related settings from config file (`answer-cookie`, `cookie-secret`, `cookie-staging-secret` and `cookie-secret-file`) will be reevaluated and effectuated after `nsd-control reconfig`

### CODE REVIEW

**Comment by wtoorop** [comment]
It's ready for review (and merge if accepted). One thing to consider/discuss is that it has one thing that is not backwards compatible, which is that cookie secrets from config file now are preferred over cookies from the cookie secrets file. Before this was the other way around. But, this feels more logical to me as zones in the configuration file are also preferred over zones in the zone list file. I believe there will be a negligible number of people that have cookie secrets in both the config file and in the cookie secrets file and would depend on the earlier preference to pick the secrets file.

**Comment by anandb-ripencc** [comment]
Looks reasonable to me. I agree that despite the backwards-incompatible change, it is unlikely to affect anyone, because no-one would define cookies in both places.

File: difffile.c
Comment (wcawijngaards) [inline difffile.c:1978]: The function `task_process_add_cookie_secret` that was deleted used `explicit_bzero`, to wipe the secrets after use. The change adds that to this function.

```suggestion
	memmove(nsd->cookie_secrets, task->zname, sizeof(nsd->cookie_secrets));
	explicit_bzero(task->zname, sizeof(nsd->cookie_secrets));
```

File: remote.c
Comment (wcawijngaards) [inline remote.c:2469]: The `backup_cookie_secrets` variable can be erased after the function is done.

```suggestion
	}
	explicit_bzero(backup_cookie_secrets, sizeof(cookie_secrets_type));
```

File: remote.c
Comment (wcawijngaards) [inline remote.c:2495]: The `backup_cookie_secrets` variable can be erased after the function is done.

```suggestion
	}
	explicit_bzero(backup_cookie_secrets, sizeof(cookie_secrets_type));
```

File: remote.c
Comment (wcawijngaards) [inline remote.c:2544]: The `backup_cookie_secrets` variable can be erased after the function is done.

```suggestion
	}
	explicit_bzero(backup_cookie_secrets, sizeof(cookie_secrets_type));
```

File: util.c
Comment (wcawijngaards) [inline util.c]: The `cookie_secrets` and `secret` variables can be erased when the function is done.

```suggestion
			explicit_bzero(cookie_secrets, sizeof(cookie_secrets));
			explicit_bzero(secret, sizeof(secret));
			return 0;
```

File: util.c
Comment (wcawijngaards) [inline util.c]: The `cookie_secrets` and `secret` variables can be erased when the function is done.

```suggestion
			explicit_bzero(cookie_secrets, sizeof(cookie_secrets));
			explicit_bzero(secret, sizeof(secret));
			return 0;
```

File: util.c
Comment (wcawijngaards) [inline util.c]: The `cookie_secrets` variable can be erased when the function is done.

```suggestion
	explicit_bzero(cookie_secrets, sizeof(cookie_secrets));
	return 1;
```

File: util.c
Comment (wcawijngaards) [inline util.c:1272]: The `secret` variable can be erased after use.

```suggestion
		explicit_bzero(secret, sizeof(secret));
		}
```

**Review by wcawijngaards — APPROVED** [review]
The code looks all right. I saw some places where the cookie secrets could possibly be wiped after temporary variables are done being used, maybe that is nice to have.

File: remote.c
Comment (k0ekk0ek) [inline remote.c:2570]: All possibilities are covered, so this is a panic situation(?)

File: util.c
Comment (k0ekk0ek) [inline util.c]: Falling back to `nsd_cookiesecrets.txt` in this function doesn't seem right. I think it's preferred to handle that in the configuration parser (or shortly after). Maybe do something like if `cookie_secret_file` is still `NULL` after parsing, set it to the previous default? That way behavior is more deterministic. Also would avoid the `cookie_secret_file` function, which seems to return `COOKIESECRETSFILE` if the `cookie_secret_file` is `NULL`. So, basically `fn` is only `NULL` if the `cookie_secret_file` is not `NULL` but empty, in which case `NULL` is returned. This seems a very convoluted way of avoiding checking whether the member is `NULL` just after parsing and setting right then(?)

File: util.c
Comment (k0ekk0ek) [inline util.c]: An `enum` is mapped to an integer, but it's better to be explicit. The order might change and this may go unnoticed. Basically, you want the branch to be taken if the secret does not originate from the configuration (`!= COOKIE_SECRETS_FROM_CONFIG`).

**Review by k0ekk0ek — APPROVED** [review]
I agree with the changes in general, just some small remarks regarding use of `enums` and one related to falling back to the previous value.

File: remote.c
Comment (wtoorop) [reply remote.c:2570]: I guess yes. Would you prefer a assert here?

Comment (wtoorop) [reply remote.c:2570]: Would you prefer an assert?

File: util.c
Comment (wtoorop) [inline util.c]:
```suggestion
	if(count && nsd->cookie_secrets_source != COOKIE_SECRETS_FROM_FILE) {
```

Comment (wtoorop) [reply util.c]: Done in [5674228](https://github.com/NLnetLabs/nsd/pull/378/commits/5674228f1cd200bbf1e5a645d55e8bbf0b90dd3d)

**Comment by anandb-ripencc** [comment]
I was reviewing the changes to the man pages and documentation, and it needs spelling and syntax fixes. Let me comment on the changes so you can fix them before merging this branch.

**Comment by wtoorop** [comment]
> I was reviewing the changes to the man pages and documentation, and it needs spelling and syntax fixes. Let me comment on the changes so you can fix them before merging this branch.

Thanks! I'm also still addressing the last comment from @k0ekk0ek, which will simplify the code somewhat and also _only does the fallback_ to the old default location for secrets (if there is nothing configured and nothing in the new default location), _when nsd just started_.

File: util.c
Comment (wtoorop) [reply util.c]: Okay, addressed in [c743742](https://github.com/NLnetLabs/nsd/pull/378/commits/c743742a6ebdc899bb6b65b0f8ff2171662517c1)

**Comment by wtoorop** [comment]
> I was reviewing the changes to the man pages and documentation, and it needs spelling and syntax fixes. Let me comment on the changes so you can fix them before merging this branch.

@anandb-ripencc I am done with review feedback and will merge once you're finished your spelling corrections.

**Comment by anandb-ripencc** [comment]
Actually, just merge it. I will then submit a PR to update the documentation. Please don't make a new release before I've submitted documentation fixes.

**Comment by k0ekk0ek** [comment]
Looks good te me @wtoorop :+1:

**Comment by wtoorop** [comment]
I accidentally committed along changes in simdzone. I'll force push the commit without the simdzone changes, and then merge.

**Pull request merged**
